### PR TITLE
Use shorter `build_key`

### LINKED ISSUE #611: Shorten hash used in environment path
### Context

Environment paths currently use a hash of the environment plus a timestamp plus the environment name. For example `99108419ad0fd922fdeff9bbc434b58d41f68e3f923a83f6a7ab19568463bc82-20230915-201619-848782-1-test` for an environment named "test". 

This path is very long (90 characters plus the environment name) and causes issues on Windows where the default filesystem limits paths to 260 characters, and some packages can have long paths. See https://github.com/conda-incubator/conda-store/issues/588. For example, `notebook` has a path with 122 characters (https://github.com/conda-incubator/conda-store/issues/588#issuecomment-1747541889), so 122 + 90 is 212 characters, leaving only 48 characters for the root of wherever the conda-store state directory is stored plus the environment name. 

Long paths can also cause issues on other platforms for some packages that are built to only be able to be installed in paths up to 255 characters. See https://github.com/conda-incubator/conda-store/issues/588#issuecomment-1751556643

### Value and/or benefit

On Windows, this can be fixed by setting a registry key, but this requires administrator privileges, so some deployments might not be able to do it. 

### Anything else?

See https://github.com/conda-incubator/conda-store/issues/588 for more technical details on the Windows path length problem. 

My suggestion would be to use a shorter hash. We can incorporate the timestamp into the hash, since it is already stored in the database. However, we would need to make sure this is done in a way that is backwards compatible.

### PR DESCRIPTION
**Pull request by nkaretnikov**
Fixes #611.

### CODE REVIEW

**Comment by nkaretnikov** [comment]
Envs work by constructing URLs based on build_key. These are then mapped to the state/storage dirs without checking whether the files are there. Active env switching is done via symlinking, which is atomic, also without checking. All this is fine. However, to make sure that access to old files is not lost via the web UI, it'll be necessary to mark old builds as such and store it in the DB for all builds before this commit (via a migration). All builds after this commit should use a new build version, which would allow distinguishing which build_key version to use for lookup.

File: conda-store-server/conda_store_server/orm.py
Comment (nkaretnikov) [inline conda-store-server/conda_store_server/orm.py:232]: This could be prefixed with `v2-` (or similar) to allow for multiple versions with 3 `-` delimiters right away. But I'm in favor of adding this in v3 or later if there are clashes. The formats are distinct now and we save 3 chars by not prefixing here, see `parse_build_key`.

Comment (nkaretnikov) [reply conda-store-server/conda_store_server/orm.py]: Nope. Cannot rely on dashes since name might contain them. Will use the hash size to differentiate between versions.

**Comment by nkaretnikov** [comment]
Seems fine. Still need to do the final review and test locally.

File: conda-store-server/conda_store_server/orm.py
Comment (jaimergp) [inline conda-store-server/conda_store_server/orm.py:244]: These lengths (4 and 16) are hardcoded and there's no explanation on why that was chosen. Instead I'd rather have a fixed-length hash à la nix.

Comment (nkaretnikov) [reply conda-store-server/conda_store_server/orm.py:244]: I like Nix hashes as well (specifically, I'm talking about nix-store hashes, since those also use the filesystem, see [here](https://nixos.org/guides/nix-pills/nix-store-paths)), but it can happen later and in a separate PR. It requires quite a bit of work and it was pointed out to me during the last team meeting that we might want to allocate our time elsewhere for now.

Here are some questions to answer for Nix hashes:
- To make this really fixed-size, we would need to design the v2 path scheme such that it also doesn't include the namespace name. It should be just: `store_directory / hash / <env contents>` where the first two components is the prefix that needs to be <= 255 to satisfy the conda prefix contraint.
- To hash like this and avoid weird issues due to mismatched version, we cannot hash the specification as we do now, we need to hash the contents of the lock file. Specs don't have the versions pinned, lock files are deterministic.
- Since the hashes are shared between users (no namespace anymore), they need to be resistant to collisions. The size of the hash and its collision resistance are related. This can be estimated to find the best size for this use case.
- Hashes are one-way (by design). But we also use `build_key`s to get the `build_id`. See `parse_build_key` and how it's used in `get_docker_image_manifest`. For that, we'd need to store an extra `hash -> id` mapping in the DB.
- IIUC, all other parts of the `build_key` are just to help with debugging or search. You can verify which env on disk corresponds to which specification, see when it was built, and get a human-readable name of the env.
- `build_id`s are unique because they are primary keys in the `Build` table. The only concern here is that they might potentially get reused by some DB backends.

How 4 and 16 were selected:
- Because only the `build_id`s matter, everything else doesn't help with the uniqueness part, just helps with debugging
- I selected 4 to still be able to verify that a spec matches the env by hashing it and have some confidence
- I selected 16 because it contained enough information to be meaningful with some env names I've tried, but also because it's short enough.

Summary:
- Since #653 adds error checking for this and documentation, it makes it easier for users to understand what's going on
- I'm going to bump 4 to 8 since it's a more standard size for truncation (even though it's not necessary here). This makes it somewhat less arbitrary. The original proposal was a unix timestamp and a truncated hash.
- I'm going to remove the truncation from names (since the linked PR has instructions for users to use shorter names if there are problems). It removes this arbitrary decision, which might lead to confusion later.
- Let's discuss the Nix-hash-like idea separately in a new issue.

Comment (jaimergp) [reply conda-store-server/conda_store_server/orm.py:247]: > It requires quite a bit of work and it was pointed out to me during the last team meeting that we might want to allocate our time elsewhere for now.

My main concern here is that we are going to break the path interface already, so maybe there's a benefit in only doing it once and for good. At the same time, if we perfect the art of migrating paths, then it's ok :) I guess it can be discussed in the weekly tomorrow.

**Comment by nkaretnikov** [comment]
Manual tests I did (as of commit 738f650):
- [x] conda-store state created on main works with changes in the PR (links to the lock file and other files work in the UI)
- [x] having build_key_version=1 in the config allows to access v2 files in the UI
- [x] having build_key_version=2 in the config allows to access v1 files in the UI
- [x] having build_key_version=3 in the config raises an error.

**Comment by nkaretnikov** [comment]
This is currently blocked due to an upstream test hanging (or taking too long?) (unrelated to this PR): https://github.com/conda-incubator/conda-store/issues/666

I'll try to fix the upstream issue first before landing this.

**Comment by nkaretnikov** [comment]
@jaimergp Updated, PTAL. Added a command-line/config parameter, now users can switch between v1 and v2 if needed. By default the short hash version is used. With the new default, old environments are still accessible in the UI.

File: conda-store-server/conda_store_server/orm.py
Comment (nkaretnikov) [reply conda-store-server/conda_store_server/orm.py:268]: There's no breakage. Everything is migrated transparently and you can switch back and forth. Nix-like hashes would be more different compared to what we have right now. I still see value in having just a truncated hash and timestamp (as in the v2 here). Nix-style hashes can be added later.

File: conda-store-server/conda_store_server/app.py
Comment (jaimergp) [inline conda-store-server/conda_store_server/app.py:387]: This feels a bit hacky. Why do we need a global constant for this?

File: conda-store-server/conda_store_server/orm.py
Comment (jaimergp) [inline conda-store-server/conda_store_server/orm.py:268]: It feels that these should be better encoded in separate callables for clarity. Something like a dictionary of versions -> callables, that implement each format cleanly. This would also double as the catalog of available versions for validation and error reporting.

File: conda-store-server/conda_store_server/orm.py
Comment (jaimergp) [inline conda-store-server/conda_store_server/orm.py:282]: This is also a bit hacky. We shouldn't rely on lossy info for this, but a well-known id->info relationship in the database. I know you know, and that that'd be considered ideal and we don't have the time but I wonder what's the rush to ship this now instead of in two weeks (still within the STF window).

If this is not going to be addressed now, and if we don't have an issue for this limitation, we should create one.

File: conda-store-server/conda_store_server/orm.py
Comment (jaimergp) [inline conda-store-server/conda_store_server/orm.py:204]: This can be a tuple.
```suggestion
        if build_key_version not in (1, 2):
```

But hardcoding these options here (separate from the definitions) is also strange. At this stage we should know which build_key_versions are available via an external name.

File: conda-store-server/conda_store_server/orm.py
Comment (jaimergp) [inline conda-store-server/conda_store_server/orm.py:205]:
```suggestion
            raise ValueError(f"invalid build_key_version={build_key_version}. Accepted values are {1, 2}")
```

Proof again that this is missing something. We should be able to easily say which ones are accepted in the error. But we would be forced to hardcode those again.

File: conda-store-server/conda_store_server/app.py
Comment (nkaretnikov) [reply conda-store-server/conda_store_server/app.py:387]: There is a big comment about it in a different file. No other way to do it due to import cycles.

```
# Avoids a cyclic dependency between the orm module and the module defining
# CondaStore.build_key_version. Because the orm module is loaded early on
# startup, we want to delay initialization of the build_key_version field until
# it's been read from the config.
```

File: conda-store-server/conda_store_server/orm.py
Comment (nkaretnikov) [reply conda-store-server/conda_store_server/orm.py:282]: I want to ship this version of the build key regardless because:
- it somewhat mitigates the length issues people might be having (see the test, the new version is 2x shorter)
- it's close enough to what we have already so risk of potential issues is minimized
- I consider Nix-like hashes a riskier change and would like to have this as an intermediate step, so people could fall back to this if needed.

File: conda-store-server/conda_store_server/__init__.py
Comment (nkaretnikov) [inline conda-store-server/conda_store_server/__init__.py:35]: Note: without `tzinfo`, this would vary between machines because `timestamp` uses the system clock by default, which might be not in UTC.

File: conda-store-server/conda_store_server/__init__.py
Comment (nkaretnikov) [inline conda-store-server/conda_store_server/__init__.py:44]: Forgot type hint here, will fix.

File: conda-store-server/conda_store_server/app.py
Comment (nkaretnikov) [inline conda-store-server/conda_store_server/app.py:121]: The current version is now set right after `build_key_version` is processed, which is better than before, when it was done only when creating a DB session.

File: conda-store-server/conda_store_server/orm.py
Comment (nkaretnikov) [inline conda-store-server/conda_store_server/orm.py:191]: All `conda_store_server` imports are local in this file to avoid a cyclic import problem mentioned in a `BuildKey` class comment.

File: conda-store-server/tests/test_actions.py
Comment (nkaretnikov) [inline conda-store-server/tests/test_actions.py:261]: This now initializes `build_key_version` via `conda_store` as one would in real life (via the config)

File: docs/administration.md
Comment (nkaretnikov) [inline docs/administration.md:621]: This also needs to be mentioned where all CondaStore options are listed in this file

**Comment by nkaretnikov** [comment]
@jaimergp Addressed your feedback, PTAL

File: conda-store-server/conda_store_server/app.py
Comment (jaimergp) [reply conda-store-server/conda_store_server/app.py:121]: Is it better to use `raise X from e` here?

File: docs/administration.md
Comment (jaimergp) [inline docs/administration.md:603]: By "environment specification" you mean the input package requests, or the solved environment?

File: conda-store-server/conda_store_server/__init__.py
Comment (jaimergp) [inline conda-store-server/conda_store_server/__init__.py:11]: Nice, this looks way better. However I still sniff some code smell here 🤔 I don't get why we are stuck with a multiversion class in the `__init__` module. I would have assumed it was simple enough to have a setup like this:

```python
# optionally, we could define a base class with the common stuff
# omitted for clarity
class BuildKeyV1:
  ...  # v1 specificities
class BuildKeyV2:
  ...  # v2 specificities, overrides as necessary

BUILD_KEY_VERSIONS = {
  1: BuildKeyV1,
  2: BuildKeyV2,
  None: BuildKeyV2,
}
```

This can be all put in a separate module if necessary. Then the consumer code will check in the versions dict which one to use according to the current config. In other words, I think we need to take a look at the problem that causes the circular import and see if there's a better way to define these ideas.

Thanks for the patience! I feel we are getting closer.

Comment (nkaretnikov) [reply conda-store-server/conda_store_server/__init__.py:11]: > I think we need to take a look at the problem that causes the circular import

What we want:
- We want a user-configurable build-key
- The argument/config parsing is done in `conda-store-server/conda_store_server/app.py:CondaStore`
- After reading a config value, we want to update the `orm` module.

Problems:
- The `orm` module is loaded earlier than `app` on startup
- The `app` module also depends on the `orm` module (imports it).

Solution:
- A common module (`__init__.py` in this case) and delayed initialization (`default=_get_build_key_version` in `orm.py`) is a way to solve this
- A common module avoids a cyclic import
- Delayed initialization avoids caching the value on early imports. It's always fetched from a common module after it's been initialized once we actually want to use the `orm` module (as opposed to just importing it).

----

> BUILD_KEY_VERSIONS

I don't understand this new proposal at all.

- What's exactly wrong with the `BuildKey` class?
- What issue does the new proposal solve compared to the existing `BuildKey` class?
- Why do we want to have 2 classes for this? We could, but I just don't see the benefit.

Moreover, I see some problems with this new proposal:
- Nothing prevents me from doing `BUILD_KEY_VERSIONS[non-existent-version]` and getting a generic error. In the current `BuildKey` implementation, error-checking is shared and hidden.
- This will still need another global variable to track the current version (to index into this dict). In the current `BuildKey` implementation, this is hidden inside the class. All `BuildKey`-related things are part of the class.
- Regarding `None: BuildKeyV2`: This is wrong and will hide potential issues if build key version is not initialized (while the user might be requesting BuildKeyV1 in the config). This can happen when a module is imported earlier than the build key version is set and the value is cached. This currently doesn't happen, but someone might run into this issue when refactoring. In my code, I have an assertion to guard against that.

**Summary**:
- I'm going to follow up offline to clarify this proposal
- Its value is not clear to me and I see some problems with it
- Implementing and testing this again would take time away from other things I could be working on.

File: conda-store-server/conda_store_server/app.py
Comment (nkaretnikov) [reply conda-store-server/conda_store_server/app.py:121]: In general, yes. But it doesn't work with traitlets. E.g.,:

```py
            raise TraitError("c.CondaStore.build_key_version") from e
```

would print

```
[CondaStoreServer] CRITICAL | Bad config encountered during initialization: c.CondaStore.build_key_version
```

No traceback is included. No additional information is printed.

**Comment by nkaretnikov** [comment]
@jaimergp PTAL. Replied to all of your comments. Only made a change to the docs.

**Review by jaimergp — APPROVED** [review]

**Pull request merged**
